modules/ipc/include/hephaestus/ipc/zenoh/action_server.h

```diff
@@ -0,0 +1,250 @@
+//================================================================================================
+// Copyright (C) 2023-2024 HEPHAESTUS Contributors
+//=================================================================================================
+
+#pragma once
+
+#include <atomic>
+#include <functional>
+#include <future>
+#include <optional>
+
+#include <magic_enum.hpp>
+
+#include "hephaestus/concurrency/message_queue_consumer.h"
+#include "hephaestus/ipc/common.h"
+#include "hephaestus/ipc/publisher.h"
+#include "hephaestus/ipc/zenoh/internal/action_server_client_helper.h"
+#include "hephaestus/ipc/zenoh/service.h"
+#include "hephaestus/ipc/zenoh/session.h"
+#include "hephaestus/ipc/zenoh/types/action_server_types.h"
+#include "hephaestus/ipc/zenoh/types/action_server_types_protobuf.h"  // NOLINT(misc-include-cleaner)
+
+namespace heph::ipc::zenoh {
+
+/// An action server is a server that execute user provided code in response to requests from clients.
+/// Upon completion a result is sent back to the client.
+/// Differently from classic request/response servers, action servers are asynchronous and non-blocking.
+/// Action servers also provide functionalities for the user to send status updates to the client and are
+/// interruptible.
+/// To instantiate an ActionServer the user needs to provide two callbacks:
+/// - RequestCallback
+///   - Takes in input the request and need to decide if the request is valid and if it can be served.
+///   - The callback should not block and return immediately,
+/// - ExecuteCallback
+///   - This is the function that does the actual work and eventually return the final response to the client.
+///   - The execute callback is executed in a dedicated thread.
+///   - The function begin execution as soon as the request is accepted.
+///   - The function has access to a Publisher to send status updates to the client.
+///     - The frequency of the updates is decided by the user; updates are not mandatory
+///   - The function can be interrupted by the client by setting the stop_requested flag to true.
+///     - The function should check this flag periodically and return if it is set.
+/// NOTES:
+/// - If an action server is already serving a request it will not accept new ones.
+/// - RequestT needs to be copyable and one copy will be made of it.
```

Comment:
right now one thing that is missing is the ability to be notified when the server has completed the request, from a process that is not the one currently waiting for the response. 
We can add such a functionality and then abstract over it.

But first I want to check if it is really needed, because if a server is called only from a single other process than the caller has a future that can use to wait for the process to terminate